impl/src/main/java/org/ehcache/internal/store/tiering/CacheStore.java

```diff
@@ -336,12 +343,37 @@ public void disableStoreEventNotifications() {
       CachingTier<K, V> cachingTier = cachingTierProvider.createCachingTier(storeConfig, serviceConfigs);
       AuthoritativeTier<K, V> authoritativeTier = authoritativeTierProvider.createAuthoritativeTier(storeConfig, serviceConfigs);
 
-      return new CacheStore<K, V>(cachingTier, authoritativeTier);
+      CacheStore<K, V> store = new CacheStore<K, V>(cachingTier, authoritativeTier);
+      registerStore(store, cachingTierProvider, authoritativeTierProvider);
+      return store;
+    }
+
+    <K, V> void registerStore(final CacheStore<K, V> store, final CachingTier.Provider cachingTierProvider, final AuthoritativeTier.Provider authoritativeTierProvider) {
+      if(providersMap.putIfAbsent(store, new AbstractMap.SimpleEntry<CachingTier.Provider, AuthoritativeTier.Provider>(cachingTierProvider, authoritativeTierProvider)) != null) {
+        throw new IllegalStateException("Instance of the Store already registered!");
+      }
     }
 
     @Override
     public void releaseStore(Store<?, ?> resource) {
-      resource.close();
+      CacheStore cacheStore = (CacheStore) resource;
```

Comment:
Oh you mean in case someone passes some wrong store in there?... so it throws `IllegalArgumentException` instead of a `ClassCastException`... got you!